src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -443,6 +458,56 @@ def _upload_assembly_files_in_parallel(
     return failed_names
 
 
+def _gen(
+        conf: Conf,
+        wait_to_upload_assemblies: dict[str, str]
+) -> Generator[tuple[tuple[str, ...], tuple[str, ...], tuple[str, ...], int], None, None]:
+    """
+    Generator function to yield the assembly files to upload.
+    """
+    assembly_files_len = len(wait_to_upload_assemblies)
+    assembly_names = tuple(wait_to_upload_assemblies.keys())
+    host_assembly_dirs = tuple(wait_to_upload_assemblies.values())
+    container_internal_assembly_paths = tuple(
+        [
+            os.path.join(JOB_DIR_IN_ASSEMBLYUTIL_CONTAINER, assembly_name)
+            for assembly_name in assembly_names
+        ]
+    )
+
+    start_idx = 0
+    cumsize = 0
+    # to avoid memory issue, upload 1GB * MEMORY_UTILIZATION of assembly files at a time
+    max_cumsize = MAX_DATA_SIZE * MEMORY_UTILIZATION
+    for idx, assembly_name in enumerate(assembly_names):
+        file_path = os.path.join(conf.job_data_dir, assembly_name)
+        if not os.path.exists(file_path):
+            raise ValueError(f"{file_path} does not exist. "
+                             f"Ensure file {assembly_name} exist in {conf.job_data_dir}")
+        file_size = os.path.getsize(file_path)
+        # cumulate assembly files until the total size is greater than max_cumsize
+        if file_size + cumsize <= max_cumsize:
+            cumsize += file_size
```

Comment:
So the 1GB limit on data uploaded to the workspace is for the data object, not any files (which are stored in the blobstore, not the workspace [and why that is is another discussion]), so this is massively undercounting the number of objects that can be safely saved in one batch since the data objects don't include any sequence data: https://ci.kbase.us/#spec/type/KBaseGenomeAnnotations.Assembly

The tricky part is that AssemblyUtil creates the objects and so only it knows the object sizes. I'd say you have two options here:

1) Update AssemblyUtil to handle the batching and parallelization. This seems like the way to go as it helps everyone.
2) Use a constant factor to calculate an estimate for the size of the assembly object based on the file size. You could use the RefSeq workspace to select, say, 10-100k random objects and get the data object size and the file size, calculate the mean & std dev, and pick a reasonable value.

Either way, I'd have a max batch size (well 10k is the max enforced by the workspace, but you could experiment to see where diminishing returns means increasing the batch size beyond, say, 100 is pointless) so that workspace connections aren't open too long.

Reply:
Hi, if my understanding is correct, this [ao](https://github.com/kbaseapps/AssemblyUtil/blob/59855bcb24014c3030e146b98c8d3e70abac1fa1/lib/AssemblyUtil/FastaToAssembly.py#L96) is the data object you are talking about.

Reply:
Right

Reply:
I feel if we put batching logic inside FastaToAssembly.py and pass in the entire wait_to_upload_assemblies dict, there will be no parallelization.

Reply:
Why do you say that?

Reply:
Let's say we set 5 workers. The first worker will take the entire wait_to_upload_assemblies dict. Updated batching logic will take care of memory issues. All other workers are stopping, and therefore no parallelization. 

Maybe we need to break that big dict into small ones and then pass into the queue as well.

Reply:
> Let's say we set 5 workers. The first worker will take the entire wait_to_upload_assemblies dict.

Why is that?

Reply:
If we don't break wait_to_upload_assemblies dict into small ones, conf.input_queue.put() will be called only once.

Reply:
You're talking about the queue in the uploader? I was thinking that queue would completely go away and we'd have all the parallelization be handled inside assemblyutils.

Maybe we should talk through the implementation in a huddle?

Reply:
Oh, now makes more sense. Let me quickly think about it before zoom/huddle.

Reply:
Addressed in AssemblyUtil

Reply:
: